File: python-solutions/knights-dialer.py
# Suppose you dial keys on the keypad using only hops a knight can make. Every time the knight lands on a key, we dial
# that key and make another hop. The starting position counts as being dialed.
#
# How many distinct numbers can you dial in N hops from a particular starting position?
# See https://hackernoon.com/google-interview-questions-deconstructed-the-knights-dialer-f780d516f029
# 6 -> 0,1,7
#   0 -> 4,6
#       4 -> 0,3,9
#       6 -> 0,1,7
#   1 -> 6,8
#       6 -> 0,1,7
#       8 -> 1,3
#   7 -> 2,6
#       2 -> 7,9
#       6 -> 0,1,7
#
# 6,0,4
# 6,0,6
# 6,1,6
# 6,1,8
# 6,7,2
# 6,7,6
#

key_map = {
    1: (6, 8),
    2: (7, 9),
    3: (4, 8),
    4: (0, 3, 9),
    5: tuple(),
    6: (0, 1, 7),
    7: (2, 6),
    8: (1, 3),
    9: (2, 4),
    0: (4, 6)
}


def neighbors(pos):
    return key_map[pos]


def yield_sequences(start_pos, n, cur_seq=None):
    if cur_seq is None:
        cur_seq = [start_pos]

    if n == 0:
        yield cur_seq
        return

    for neighbor in neighbors(start_pos):
        yield from yield_sequences(neighbor, n - 1, cur_seq + [neighbor])


def yield_seq_iter(start_pos, n):
    results = [[start_pos]]

    while n > 0:
        cur_list = []

        for num_list in results:
            # get the next neighbors for each list of numbers
            nn = neighbors(num_list[-1])

            for next_num in nn:
                cur_seq = num_list.copy()
                cur_seq.append(next_num)
                cur_list.append(cur_seq)

        results = cur_list
        n -= 1
    return results


def count_sequences_recursive(start_position, n):
    if n == 0:
        return 1

    num_sequences = 0
    for position in neighbors(start_position):
        num_sequences += count_sequences_recursive(position, n - 1)
    return num_sequences


def count_sequences_iter(start_position, n):
    num_sequences = 0
    sequences = [start_position]
    while n > 0:
        next_seq = []

        for pos in sequences:
            nn = neighbors(pos)
            next_seq += nn

        num_sequences = len(next_seq)
        sequences = next_seq
        n -= 1
    return num_sequences


# for seq in yield_seq_iter(6, 3):
#     print(seq)

# print("----------------")
# for seq in yield_sequences(6, 2):
#     print(seq)
print(count_sequences_iter(6, 20))
# print(count_sequences_recursive(6, 20))

File: rust-solutions/google/src/bin/knights-dialer.rs
// Suppose you dial keys on the keypad using only hops a knight can make. Every time the knight
// lands on a key, we dial that key and make another hop. The starting position counts as
// being dialed.
// How many distinct numbers can you dial in N hops from a particular starting position?
// See https://hackernoon.com/google-interview-questions-deconstructed-the-knights-dialer-f780d516f029
// 6 -> 0,1,7
//   0 -> 4,6
//       4 -> 0,3,9
//       6 -> 0,1,7
//   1 -> 6,8
//       6 -> 0,1,7
//       8 -> 1,3
//   7 -> 2,6
//       2 -> 7,9
//       6 -> 0,1,7
// 6,0,4
// 6,0,6
// 6,1,6
// 6,1,8
// 6,7,2
// 6,7,6

use once_cell::sync::Lazy;
use std::collections::HashMap;

static KEY_MAP: Lazy<HashMap<u8, Vec<u8>>> = Lazy::new(|| {
    let mut hm = HashMap::with_capacity(10);
    hm.insert(1, vec![6, 8]);
    hm.insert(2, vec![7 ,9]);
    hm.insert(3, vec![4 ,8]);
    hm.insert(4, vec![0, 3, 9]);
    hm.insert(5, vec![]);
    hm.insert(6, vec![0, 1, 7]);
    hm.insert(7, vec![2, 6]);
    hm.insert(8, vec![1, 3]);
    hm.insert(9, vec![2 ,4]);
    hm.insert(0, vec![4 ,6]);

    hm
});


fn neighbors(key: u8) -> &'static Vec<u8> {
    KEY_MAP.get(&key).expect("key must be in the range 0 - 9")
}

fn yield_seq_iter(start_pos: u8, n: usize) -> Vec<Vec<u8>> {
    let mut results = vec![vec![start_pos]];
    let mut n = n;

    while n > 0 {
        let mut cur_list: Vec<Vec<u8>> = Vec::new();

        for num_list in &results {
            let nn = neighbors(*num_list.last().unwrap());
            
            for n in nn {
                let mut cur_seq = num_list.clone();
                cur_seq.push(*n);
                cur_list.push(cur_seq);
            }
        }
        results = cur_list;
        n -= 1;
    }

    results
}

fn main() {
    let seqs = yield_seq_iter(6, 2);
    for seq in seqs.iter() {
        println!("{:?}", seq);
    }
}